Wobble/Scheduling/TaskHandler.cs

```diff
@@ -0,0 +1,119 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Wobble.Logging;
+
+namespace Quaver.Shared.Scheduling
+{
+    /// <summary>
+    ///     This will be used to handle threaded tasks.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <typeparam name="TResult"></typeparam>
+    public class TaskHandler<T, TResult> : IDisposable
+    {
+        private Func<T, TResult> Function { get; }
+
+        /// <summary>
+        ///     Event Invoked when the task gets completed.
+        /// </summary>
+        public event EventHandler<TaskCompleteEventArgs<T, TResult>> OnCompleted;
+
+        /// <summary>
+        ///     Event Invoked when the task gets cancelled.
+        /// </summary>
+        public event EventHandler<TaskCancelledEventArgs<T>> OnCancelled;
+
+        /// <summary>
+        ///     Used to handle Task Cancellation.
+        /// </summary>
+        public CancellationTokenSource Source { get; private set; } = new CancellationTokenSource();
+
+        /// <summary>
+        ///     Determined by if current task is completed.
+        /// </summary>
+        public bool IsCompleted { get; private set; }
+
+        /// <summary>
+        ///     Determined by if the current task is running.
+        /// </summary>
+        public bool IsRunning { get; private set; }
+
+        /// <summary>
+        ///     Determined by if the current task is cancelled.
+        /// </summary>
+        public bool IsCancelled { get; private set; }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="function"></param>
+        public TaskHandler(Func<T, TResult> function) => Function = function;
+
+        /// <summary>
+        ///     This method will cancel the previous task and run a new task.
+        /// </summary>
+        /// <param name="input"></param>
+        /// <param name="delay"></param>
+        public void Run(T input, int delay = 0)
+        {
+            // We want to automatically cancel the previous task and dispose the other
+            // cancellation token source before running a new task.
+            Source.Cancel();
```

Comment:
Right after this line there should be some kind of synchronization to wait for the running thread to finish. Otherwise there is a race condition: you set `IsCancelled = false` right below, but the cancelled thread will set `IsCancelled = true` in its handler.

Reply:
Actually, the intention here is to cancel the running task without blocking, so it's not desirable to do any synchronization. This needs to be handled in some other way then. Here's a plausible scenario:
1. `taskHandler.Run();` is called, it's in the middle of a long `Function();` in its thread.
2. `taskHandler.Run();` is called again, the token is cancelled, and it starts running a second thread.
3. The first thread finishes its `Function();`, the next line throws the cancelled exception, the handler is called, and `taskHandler` is now marked `IsCancelled = true`, even though the second thread is currently running fine.